x/bank/keeper.go

```diff
@@ -73,20 +69,119 @@ func (ck CoinKeeper) SendCoins(ctx sdk.Context, fromAddr sdk.Address, toAddr sdk
+	am sdk.AccountMapper
+}
+
+// NewCoinKeeper returns a new CoinKeeper
+func NewCoinKeeper(am sdk.AccountMapper) CoinKeeper {
+	return CoinKeeper{am: am}
+}
+
+// GetCoins returns the coins at the addr.
+func (keeper CoinKeeper) GetCoins(ctx sdk.Context, addr sdk.Address) sdk.Coins {
+	return getCoins(ctx, keeper.am, addr)
+}
+
+// SetCoins sets the coins at the addr.
+func (keeper CoinKeeper) SetCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) sdk.Error {
+	return setCoins(ctx, keeper.am, addr, amt)
+}
+
+// HasCoins returns whether or not an account has at least amt coins.
+func (keeper CoinKeeper) HasCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) bool {
+	return hasCoins(ctx, keeper.am, addr, amt)
+}
+
+// SubtractCoins subtracts amt from the coins at the addr.
+func (keeper CoinKeeper) SubtractCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
+	return subtractCoins(ctx, keeper.am, addr, amt)
+}
+
+// AddCoins adds amt to the coins at the addr.
+func (keeper CoinKeeper) AddCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
+	return addCoins(ctx, keeper.am, addr, amt)
+}
+
+// SendCoins moves coins from one account to another
+func (keeper CoinKeeper) SendCoins(ctx sdk.Context, fromAddr sdk.Address, toAddr sdk.Address, amt sdk.Coins) sdk.Error {
+	return sendCoins(ctx, keeper.am, fromAddr, toAddr, amt)
+}
+
+// InputOutputCoins handles a list of inputs and outputs
+func (keeper CoinKeeper) InputOutputCoins(ctx sdk.Context, inputs []Input, outputs []Output) sdk.Error {
+	return inputOutputCoins(ctx, keeper.am, inputs, outputs)
+}
+
+// --------------------------------------------------
+
+// SendKeeper only allows transfers between accounts, without the possibility of creating coins
+type SendKeeper struct {
+	am sdk.AccountMapper
+}
+
+// NewSendKeeper returns a new CoinKeeper
+func NewSendKeeper(am sdk.AccountMapper) SendKeeper {
+	return SendKeeper{am: am}
+}
+
+// GetCoins returns the coins at the addr.
+func (keeper SendKeeper) GetCoins(ctx sdk.Context, addr sdk.Address) sdk.Coins {
+	return getCoins(ctx, keeper.am, addr)
+}
+
+// HasCoins returns whether or not an account has at least amt coins.
+func (keeper SendKeeper) HasCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) bool {
+	return hasCoins(ctx, keeper.am, addr, amt)
+}
+
+// SendCoins moves coins from one account to another
+func (keeper SendKeeper) SendCoins(ctx sdk.Context, fromAddr sdk.Address, toAddr sdk.Address, amt sdk.Coins) sdk.Error {
+	return sendCoins(ctx, keeper.am, fromAddr, toAddr, amt)
+}
+
+// InputOutputCoins handles a list of inputs and outputs
+func (keeper SendKeeper) InputOutputCoins(ctx sdk.Context, inputs []Input, outputs []Output) sdk.Error {
+	return inputOutputCoins(ctx, keeper.am, inputs, outputs)
+}
+
+// --------------------------------------------------
+
+// ViewKeeper only allows reading of balances
+type ViewKeeper struct {
+	am sdk.AccountMapper
```

Comment:
Do we actually need to create explicit structs for this? Is it not safe for some reason to use just create more limited interfaces and of CoinKeeper and cast a new object to that interface? I'm actually not sure and would like to know

Reply:
More on this I really like @cwgoes design https://github.com/cosmos/cosmos-sdk/issues/802#issuecomment-380065139

Reply:
base on 802 conversation ... YES we do want explicit structs - aka this comment is stale